### PR TITLE
new multicluster handler

### PR DESCRIPTION
**Pull request by howardjohn**
This PR refactors the multicluster logic.

Before, a multicluster aware controller would implement an interface like so:

```go
type ClusterHandler interface {
	ClusterAdded(cluster *Cluster, stop <-chan struct{})
	ClusterUpdated(cluster *Cluster, stop <-chan struct{})
	ClusterDeleted(clusterID cluster.ID)
}
```

We currently have 3 controllers implementing this, and will have a 4th. Additionally, one of them is a giant monolith controller that does way more than it should -- ideally its broken down, which would make this problem worse (without this PR). Each controller needs to manage state themselves, which ends up being a lot of code.

After, a multicluster aware controller registers a `f func(cluster *Cluster, stop <-chan struct{}) T`, where T is some controller type. In return, they get back a type with

```go
ForCluster(clusterID cluster.ID) *T
All() []T
```

State of the clusters is managed automatically. When a cluster is deleted, `T.close()` is called, allowing for cleanup.

---

This PR is a WIP. It compiles but I didn't move any of the tests over, clean it up, etc. Wanted to get some feedback first.

### CODE REVIEW

**Comment by howardjohn** [comment]
cc @ramaraochavali @hzxuzhonghu @stevenctl for opinions

File: pkg/kube/multicluster/secretcontroller.go [hunk 1]
```diff
@@ -142,7 +200,14 @@ func NewController(kubeclientset kube.Client, namespace string, clusterID cluste
 	return controller
 }
 
-func (c *Controller) AddHandler(h ClusterHandler) {
+func NewHandler[T Closer](f func(cluster *Cluster, stop <-chan struct{}) T) *MultiCluster[T] {
```
Comment (ramaraochavali) [inline pkg/kube/multicluster/secretcontroller.go:203]: May be we can combine NewHandler and RegisterHander?

File: pkg/kube/multicluster/secretcontroller.go [hunk 2]
```diff
@@ -60,10 +62,66 @@ var (
 	remoteClusters = clustersCount.With(clusterType.Value("remote"))
 )
 
-type ClusterHandler interface {
-	ClusterAdded(cluster *Cluster, stop <-chan struct{})
-	ClusterUpdated(cluster *Cluster, stop <-chan struct{})
-	ClusterDeleted(clusterID cluster.ID)
+type Closer interface {
```
Comment (ramaraochavali) [inline pkg/kube/multicluster/secretcontroller.go:65]: nit: Closeable?

File: pkg/kube/multicluster/secretcontroller.go (hunk 1 above)
Comment (howardjohn) [reply pkg/kube/multicluster/secretcontroller.go:203]: I would but you cannot have generic methods in go unfortunately...

Or we could make it not use generics and use any instead

Comment (ramaraochavali) [reply pkg/kube/multicluster/secretcontroller.go:203]: I see. Trying to see if we can avoid two calls

Comment (howardjohn) [reply pkg/kube/multicluster/secretcontroller.go:203]: it can be one if it takes Controller as an argument. probably that's better

Comment (ramaraochavali) [reply pkg/kube/multicluster/secretcontroller.go:203]: Yes. That sounds better

File: pkg/kube/multicluster/secretcontroller.go (hunk 2 above)
Comment (saiskee) [reply pkg/kube/multicluster/secretcontroller.go:65]: https://go.dev/doc/effective_go#interface_names

> By convention, one-method interfaces are named by the method name plus an -er suffix

File: pkg/kube/multicluster/secretcontroller.go [hunk 3]
```diff
@@ -142,7 +151,24 @@ func NewController(kubeclientset kube.Client, namespace string, clusterID cluste
...
+func BuildMultiClusterComponent[T ComponentConstraint](c ComponentBuilder, constructor func(cluster *Cluster) T) *Component[T] {
+	comp := &Component[T]{
+		constructor: constructor,
+		clusters:    make(map[cluster.ID]T),
+	}
+	c.registerHandler(comp)
+	return comp
+}
+
+func (c *Controller) registerHandler(h handler) {
```
Comment (hzxuzhonghu) [inline pkg/kube/multicluster/secretcontroller.go:171]: lock handlers?

File: pilot/pkg/credentials/kube/multicluster.go [hunk 4]
```diff
@@ -16,92 +16,54 @@ package kube
...
-func NewMulticluster(configCluster cluster.ID) *Multicluster {
+func NewMulticluster(configCluster cluster.ID, controller multicluster.ComponentBuilder) *Multicluster {
 	m := &Multicluster{
-		remoteKubeControllers: map[cluster.ID]*CredentialsController{},
-		configCluster:         configCluster,
+		configCluster: configCluster,
 	}
 
+	m.component = multicluster.BuildMultiClusterComponent(controller, func(cluster *multicluster.Cluster) *CredentialsController {
+		return NewCredentialsController(cluster.Client, m.secretHandlers)
```
Comment (hzxuzhonghu) [inline pilot/pkg/credentials/kube/multicluster.go:40]: secretHandlers are registered from `(m *Multicluster) AddSecretHandler` later, how can that be passed during the initiation

Comment (stevenctl) [reply pilot/pkg/credentials/kube/multicluster.go:40]: Pre-existing ones are added here, then new ones get added later during `AddSecretHandler`.

File: pilot/pkg/credentials/kube/multicluster.go [hunk 5]
```diff
@@ -16,92 +16,54 @@ package kube
...
+	if cc := m.component.ForCluster(m.configCluster); cc != nil {
+		agg.controllers = append(agg.controllers, *cc)
 	}
-	agg.controllers = append(agg.controllers, m.remoteKubeControllers[m.configCluster])
 	return agg, nil
 }
 
 func (m *Multicluster) AddSecretHandler(h func(name string, namespace string)) {
 	m.secretHandlers = append(m.secretHandlers, h)
-	m.m.Lock()
```
Comment (stevenctl) [inline pilot/pkg/credentials/kube/multicluster.go:100]: Oh, I see what @hzxuzhonghu was saying. SecretHandler needs to either be a shared reference OR we iterate over `m.component.All()` here.

**Comment by stevenctl** [comment]
General idea +1

File: pkg/kube/multicluster/secretcontroller.go (hunk 3 above)
Comment (howardjohn) [reply pkg/kube/multicluster/secretcontroller.go:171]: This is intended and matches the existing behavior -- I will add a comment about why

File: pkg/kube/multicluster/component.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	return comp
+}
+
+func (m *Component[T]) clusterUpdated(cluster *Cluster) {
+	comp := m.constructor(cluster)
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// If there is an old one, close it
+	if old, f := m.clusters[cluster.ID]; f {
+		old.Close()
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/multicluster/component.go:65]: Good call, done

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 7]
```diff
@@ -48,11 +46,27 @@ const (
...
-var _ multicluster.ClusterHandler = &Multicluster{}
-
 type kubeController struct {
+	MeshServiceController *aggregate.Controller
 	*Controller
 	workloadEntryController *serviceentry.Controller
+	stop                    chan struct{}
+}
+
+func (k *kubeController) Close() {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:56]: This comes from https://github.com/istio/istio/blob/2ae746e168a098f49f7b6fcc1e6ee352ac7f52a6/pilot/pkg/serviceregistry/kube/controller/multicluster.go#L392-L404

File: pilot/pkg/credentials/kube/secrets.go [hunk 8]
```diff
@@ -291,10 +307,3 @@ func extractRoot(scrt *v1.Secret) (certInfo *credentials.CertInfo, err error) {
 	return nil, fmt.Errorf("found secret, but didn't have expected keys %s or %s; found: %s",
 		GenericScrtCaCert, TLSSecretCaCert, found)
 }
-
-func (s *CredentialsController) AddEventHandler(h func(name string, namespace string)) {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/credentials/kube/secrets.go:295]: This never supported adding handlers after it started (even in the old way). So now AddSecretHandler builds up `m.secretHandlers` which are passed one time to NewCredentialsController.

File: pilot/pkg/credentials/kube/multicluster.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/credentials/kube/multicluster.go:40]: See https://github.com/istio/istio/pull/49260#discussion_r1500073199

Comment (howardjohn) [reply pilot/pkg/credentials/kube/multicluster.go:40]: Resolving to consolidate conversation on https://github.com/istio/istio/pull/49260#discussion_r1500073199

File: pilot/pkg/credentials/kube/multicluster.go (hunk 5 above)
Comment (howardjohn) [reply pilot/pkg/credentials/kube/multicluster.go:100]: https://github.com/istio/istio/pull/49260#discussion_r1500073199

**Review by hzxuzhonghu — APPROVED** [review]

**Review by stevenctl — APPROVED** [review]

**Pull request merged**